Compilation with `USE_MPI=ON` fails with gcc8 (and probably newer gccs too)

Hey,

There is one obscure warning generated when compiling MPI with newer GCCs:
```

[ 39%] Building CXX object src/CMakeFiles/marian.dir/training/graph_group_async.cpp.o
[ 39%] Building CXX object src/CMakeFiles/marian.dir/training/graph_group_sync.cpp.o
In file included from /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/mpicxx.h:283,
                 from /usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h:2704,
                 from /home/nbogoych/marian-dev-onednn/src/training/communicator.h:14,
                 from /home/nbogoych/marian-dev-onednn/src/training/scheduler.h:7,
                 from /home/nbogoych/marian-dev-onednn/src/training/graph_group.h:9,
                 from /home/nbogoych/marian-dev-onednn/src/training/graph_group_async.h:5,
                 from /home/nbogoych/marian-dev-onednn/src/training/graph_group_async.cpp:1:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h: In member function ‘virtual void MPI::Op::Init(void (*)(const void*, void*, int, const MPI::Datatype&), bool)’:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:121:46: error: cast between incompatible function types from ‘void (*)(void*, void*, int*, ompi_datatype_t**, void (*)(void*, void*, int*, ompi_datatype_t**))’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Werror=cast-function-type]
     (void)MPI_Op_create((MPI_User_function*) ompi_mpi_cxx_op_intercept,
                                              ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:123:59: error: cast between incompatible function types from ‘void (*)(const void*, void*, int, const MPI::Datatype&)’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Werror=cast-function-type]
     ompi_op_set_cxx_callback(mpi_op, (MPI_User_function*) func);
                                                           ^~~~
In file included from /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/mpicxx.h:283,
                 from /usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h:2704,
                 from /home/nbogoych/marian-dev-onednn/src/training/communicator.h:14,
                 from /home/nbogoych/marian-dev-onednn/src/training/scheduler.h:7,
                 from /home/nbogoych/marian-dev-onednn/src/training/graph_group.h:9,
                 from /home/nbogoych/marian-dev-onednn/src/training/graph_group_sync.h:3,
                 from /home/nbogoych/marian-dev-onednn/src/training/graph_group_sync.cpp:1:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h: In member function ‘virtual void MPI::Op::Init(void (*)(const void*, void*, int, const MPI::Datatype&), bool)’:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:121:46: error: cast between incompatible function types from ‘void (*)(void*, void*, int*, ompi_datatype_t**, void (*)(void*, void*, int*, ompi_datatype_t**))’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Werror=cast-function-type]
     (void)MPI_Op_create((MPI_User_function*) ompi_mpi_cxx_op_intercept,
                                              ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:123:59: error: cast between incompatible function types from ‘void (*)(const void*, void*, int, const MPI::Datatype&)’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Werror=cast-function-type]
     ompi_op_set_cxx_callback(mpi_op, (MPI_User_function*) func);
                                                           ^~~~
At global scope:
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
```
After some googling it seems to be a [false positive](https://github.com/open-mpi/ompi/issues/5157) due to MPI's legacy.

There are couple of possible fixes, either disable `-Wextra`, or suggestion to add an undocumented macro:
```
#define OMPI_SKIP_MPICXX 1
#include <mpi.h>
```

Cheers,

Nick

--- a/src/training/communicator.h
+++ b/src/training/communicator.h
@@ -11,6 +11,7 @@
 #pragma GCC diagnostic ignored "-Wsuggest-override"
 #endif
 #undef HOST
+#define OMPI_SKIP_MPICXX 1 // Fixes compilation with GCC8+ https://github.com/open-mpi/ompi/issues/5157
 #include "mpi.h"
 #ifdef __GNUC__
 #pragma GCC diagnostic pop

